File: 0509-inorder-successor-in-bst-ii/0509-inorder-successor-in-bst-ii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def inorderSuccessor(self, node: 'Node') -> 'Optional[Node]':

        def goleft(root):
            if not root.left:
                return root
            return goleft(root.left)
        
        def goup(root):
            if not root:
                return None
            if node.val < root.val:
                return root
            return goup(root.parent)

        
        if node.right:
            return goleft(node.right)
        return goup(node.parent)

File: 0509-inorder-successor-in-bst-ii/0509-inorder-successor-in-bst-ii.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
};
*/

class Solution {
    public Node inorderSuccessor(Node node) {
        if (node.right != null) {
            return goLeft(node.right);
        }
        return goUp(node, node.val);
    }

    private Node goLeft(Node root) {
        if (root.left == null) {
            return root;
        }
        return goLeft(root.left);
    }
    
    private Node goUp(Node root, int target) {
        if (root == null) {
            return root;
        }
        if (target < root.val) {
            return root;
        }
        return goUp(root.parent, target);
    }
}